nni/nas/oneshot/pytorch/supermodule/operation.py

```diff
@@ -673,34 +701,87 @@ def forward_with_args(
+
+        for name in ["out_proj.weight", "out_proj.bias"]:
+            value = params_mapping.get(name, None)
+            if value is None:
+                continue
+            destination[prefix + name] = value if keep_vars else value.detach()
+
+        for name, module in self._modules.items():
+            if isinstance(module, nn.modules.linear.NonDynamicallyQuantizableLinear):
+                continue
+            if module is not None:
+                sub_state_dict(module, destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
 
-            # The rest part is basically same as pytorch
-            attn_output, attn_output_weights = F.multi_head_attention_forward(
-                query, key, value, used_embed_dim, num_heads,
-                cast(Tensor, in_proj_weight), cast(Tensor, in_proj_bias),
-                bias_k, bias_v, self.add_zero_attn,
-                dropout, out_proj_weight, cast(Tensor, out_proj_bias),
-                training=self.training,
-                key_padding_mask=key_padding_mask, need_weights=need_weights,
-                attn_mask=attn_mask, use_separate_proj_weight=True,
-                q_proj_weight=q_proj, k_proj_weight=k_proj, v_proj_weight=v_proj)
+    def forward_with_args(
+        self,
+        embed_dim: int_or_int_dict, num_heads: int,
+        kdim: int_or_int_dict | None, vdim: int_or_int_dict | None,
+        dropout: float,
+        query: torch.Tensor, key: torch.Tensor, value: torch.Tensor,
+        key_padding_mask: torch.Tensor | None = None,
+        need_weights: bool = True, attn_mask: torch.Tensor | None = None
+    ) -> tuple[torch.Tensor, torch.Tensor | None]:
+
+        if any(isinstance(arg, dict) for arg in [num_heads, dropout]):
+            raise ValueError(_diff_not_compatible_error.format('num_heads and dropout', 'MultiHeadAttention'))
+
+        if getattr(self, 'batch_first', False):
+            # for backward compatibility: v1.7 doesn't have batch_first
+            query, key, value = [x.transpose(1, 0) for x in (query, key, value)]
+
+        if isinstance(embed_dim, dict):
+            used_embed_dim = self.embed_dim
         else:
-            # Cast tensor here because of a bug in pytorch stub
-            attn_output, attn_output_weights = F.multi_head_attention_forward(
-                query, key, value, used_embed_dim, num_heads,
-                cast(Tensor, in_proj_weight), cast(Tensor, in_proj_bias),
-                bias_k, bias_v, self.add_zero_attn,
-                dropout, out_proj_weight, cast(Tensor, out_proj_bias),
-                training=self.training,
-                key_padding_mask=key_padding_mask, need_weights=need_weights,
-                attn_mask=attn_mask)
+            used_embed_dim = embed_dim
+
+        in_proj_bias, in_proj_weight, bias_k, bias_v, \
+            out_proj_weight, out_proj_bias, q_proj, k_proj, v_proj, qkv_same_embed_dim = self._slice_param(embed_dim, kdim, vdim)
```

Comment:
Why do you need a `_slice_param`?

Seems that you can directly call `_slice_param_mapping`.